R/S3_predict.R

```diff
@@ -184,20 +184,75 @@ predict.iv_robust <- function(object,
                               na.action = na.pass,
                               ...) {
 
+  X <- get_X(object, newdata, na.action)
+
+  coefs <- as.matrix(coef(object))
+
+  beta_na <- is.na(coefs[, 1])
+
+  # Get predicted values
+  preds <- X[, !beta_na, drop = FALSE] %*% coefs[!beta_na, ]
+  if (object[["fes"]]) {
+    preds <- add_fes(preds, object, newdata)
+  }
+  return(drop(preds))
+}
+
+get_X <- function(object, newdata, na.action) {
   # Get model matrix
-  rhs_terms <- delete.response(object$terms_regressors)
-  mf <- model.frame(rhs_terms, newdata, na.action = na.action)
+  if (is.null(object[["terms_regressors"]])) {
+    rhs_terms <- delete.response(object[["terms"]])
+  } else {
+    rhs_terms <- delete.response(object[["terms_regressors"]])
+  }
+  mf <- model.frame(rhs_terms, newdata, na.action = na.action, xlev = object[["xlevels"]])
 
   # Check class of columns in newdata match those in model fit
   if (!is.null(cl <- attr(rhs_terms, "dataClasses"))) .checkMFClasses(cl, mf)
 
+  if (object[["fes"]]) {
+    attr(rhs_terms, "intercept") <- 0
+  }
+
   X <- model.matrix(rhs_terms, mf, contrasts.arg = object$contrasts)
 
-  coefs <- as.matrix(coef(object))
+  return(X)
+}
 
-  beta_na <- is.na(coefs[, 1])
+add_fes <- function(preds, object, newdata) {
 
-  # Get predicted values
-  predictor <- drop(X[, !beta_na, drop = FALSE] %*% coefs[!beta_na, ])
-  return(predictor)
+  # Add factors!
+  args <- as.list(object[["call"]])
+
+  if (length(all.vars(rlang::f_rhs(args[["fixed_effects"]]))) > 1) {
+    stop(
+      "Can't use `predict.lm_robust` with more than one set of ",
+      "`fixed_effects`. Can recover fits in `fitted.values` in the model ",
+      "object."
+    )
+  }
+
+  # cast as factor
+  femat <- model.matrix(
+    ~ 0 + .,
+    data = as.data.frame(
+      sapply(
+        stats::model.frame.default(
+          args[["fixed_effects"]],
+          data = newdata,
+          na.action = NULL
+        ),
+        function(fe) as.factor(fe)
```

Comment:
as.factor

Reply:
consider writing a helper function, I've seen this block before.

Reply:
also use lapply instead of sapply, you don't want an extra simplify called accidentally.

Reply:
Yes you've seen it before, but the other one is evaluating a quosure inside the apply.